### PR TITLE
Add tests for extended_properties feature

### LINKED ISSUE #108: Add test for extraProperties
**Summary**
You can specify extra properties available for authorization and use this to normalize values across multiple auth providers. e.g two OIDC providers who have claims at different locations.

**Documentation**
https://github.com/Kuadrant/authorino/blob/main/docs/features.md#extra-identity-extension-extendedproperties
https://github.com/Kuadrant/authorino/blob/main/docs/user-guides/token-normalization.md

**Test cases**
- [ ] extraProperties
- [ ] Token normalization scenario
- [ ] https://github.com/Kuadrant/authorino/issues/395

### PR DESCRIPTION
**Pull request by azgabur**
- [x] `extended_properties` functionality implementation (contains fix of `Value` class)
- [x] Test cases:
    - [x] Token normalization
    - [x] Other extended_properties tests
    - [x] Overwrite tests
 
Contains also:

-  bugfix of "Add plain identity" function https://github.com/Kuadrant/testsuite/pull/212#discussion_r1272674382
- Replace Union with more native type marking https://github.com/Kuadrant/testsuite/pull/212#discussion_r1281998698
- ~Replace custom `asdict()` in `ValueFrom` with dict transformation https://github.com/Kuadrant/testsuite/pull/212#discussion_r1282710508~ 
- Fix typo in docstring for `asdict()` https://github.com/Kuadrant/testsuite/pull/212#discussion_r1285922191
- Change `add_item` function to force keyword-only common_features arguments https://github.com/Kuadrant/testsuite/pull/212#discussion_r1305881457

Closes #108

### CODE REVIEW

**Comment by azgabur** [comment]
^Rebase, black formatting, finish Token normalization test

**Comment by azgabur** [comment]
Rebase, add new overwrite tests

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    """
+    This test checks extended properties function for sections in identity providers.
+    Currently test checks these identity sections:
+    - Anonymous
+    - API key
+    - Oidc
+
+    EXTENDED_PROPERTIES list contains testing properties added to every identity section.
+    """
+    authorization.identity.remove_all()  # clear before test
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py:59]: The structure of this test seems weird:
1) ideally, you would not need to make changes to authorization after the setup
2) Seems like an ideal candidate for parametrization

File: testsuite/objects/__init__.py [hunk 2]
```diff
@@ -63,6 +63,33 @@ def to_dict(self):
         return {"ttl": self.ttl, "key": self.key.to_dict()}
 
 
+@dataclass
+class JsonProperty:
```
Comment (averevki) [inline testsuite/objects/__init__.py:67]: Won't be better to inherit a `Value` class here and add a `name` property instead of creating dataclass? I believe directly creating `JsonProperty` with name and value would be easier than importing both `Value` and `JsonProperty` and creating one dataclass from them. WDYT?

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+def test_overwrite(client, authorization, module_label, auth_api_key):
+    """
+    This test checks extendedProperty overwrite feature.
+    API_KEY_VALUE_B64 contains api key value in base64 as it is stored that way in AuthJson.
+    If overwrite in extendedProperty is set to False the auth.identity.type will have value of API_KEY_VALUE_B64
+    If overwrite in extendedProperty is set to True the auth.identity.type will be overwritten to value "rewritten"
+    """
+    rewrite = "rewritten"
+    for overwrite, value in [(False, API_KEY_VALUE_B64), (True, rewrite)]:
```
[the remark this answers is not in the document]
Comment (azgabur) [reply testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py]: I parametrized :)

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py (hunk 1 above)
Comment (azgabur) [reply testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py]: Parametrized as well

File: testsuite/openshift/objects/auth_config/sections.py [hunk 4]
```diff
@@ -54,6 +64,23 @@ def add_item(
...
+        metrics: bool = None,
+        cache: Cache = None,
+        extended_properties: list[ExtendedProperty] = None,
+    ):
+        """
+        Adds optional extendedProperties feature specific to IdentitySection and then calls parent add_idem() method
+        """
+        if extended_properties:
+            value["extendedProperties"] = [i.to_dict() for i in extended_properties]
+        super().add_item(name, value, priority, when, metrics, cache)
```
[the remark this answers is not in the document]
Comment (azgabur) [reply testsuite/openshift/objects/auth_config/sections.py:82]: Due to `overwrite` variable needing to be last I need to explicitly list all variables of overwritten `add_idem` function.

File: testsuite/objects/__init__.py (hunk 2 above)
Comment (azgabur) [reply testsuite/objects/__init__.py]: Refactored the data classes

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+def auth_api_key(create_api_key, module_label):
+    """Creates API key Secret and returns auth for it."""
+    api_key = create_api_key("api-key", module_label, API_KEY_VALUE)
+    return HeaderApiKeyAuth(api_key)
+
+
+@pytest.fixture(scope="module")
+def authorization(authorization, module_label, rhsso, request):
+    """Add simple response to inspect 'auth.identity' part of authJson."""
+    if request.param == "anonymous":
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py:38]: This IF is wrong, very, very wrong :)

**Comment by azgabur** [comment]
Comments/Documentation still needs to be rewritten.

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    """
+    This test checks extended properties function for sections in identity providers.
+    Currently test checks these identity sections:
+    - Anonymous
+    - API key
+    - Oidc
+
+    EXTENDED_PROPERTIES list contains testing properties added to every identity section.
+    """
+    del authorization  # unused
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py:88]: why? Do not request authorization then

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    assert extract_response(response)["property_rewrite_false"] % "" == "original"
+    assert extract_response(response)["property_chain_static"] % "" is None
+    assert extract_response(response)["property_chain_dynamic"] % "" is None
+    assert extract_response(response)["property_chain_self"] % "" is None
+
+
+@pytest.fixture(scope="module")
+def expected_value(request):
+    """abc"""
+    if request.param == "overwrite_false":
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py:103]: No if please

File: testsuite/objects/__init__.py [hunk 8]
```diff
@@ -63,6 +65,41 @@ def to_dict(self):
...
+        return {"name": self.name, **super().to_dict()}
+
+
+@dataclass
+class ExtendedProperty(JsonProperty):
+    """
+    Class defining ExtendedProperty which has more features than regular JsonProperty
+    """
+
+    __slots__ = ("overwrite",)
```
Comment (pehala) [inline testsuite/objects/__init__.py:92]: Why? You can extend dataclasses and specify additional parameters as usual

Comment (azgabur) [reply testsuite/objects/__init__.py]: Addressed

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (azgabur) [reply testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py]: No longer applicable

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py (hunk 1 above)
Comment (azgabur) [reply testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py]: No longer applicable

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py (hunk 5 above)
Comment (azgabur) [reply testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py]: No longer applicable

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py (hunk 6 above)
Comment (azgabur) [reply testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py]: No longer applicable

File: testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py (hunk 7 above)
Comment (azgabur) [reply testsuite/tests/kuadrant/authorino/identity/test_extended_properties.py]: No longer applicable

**Comment by azgabur** [comment]
Re-did basic extended properties test. Rebased and squashed commits. All tasks complete. Ready for review.

File: testsuite/openshift/objects/auth_config/sections.py [hunk 9]
```diff
@@ -130,7 +157,7 @@ def anonymous(self, name, **common_features):
     @modify
     def plain(self, name, auth_json, **common_features):
         """Adds plain identity"""
-        self.add_item(name, {"plain": {"authJSON": auth_json}, **common_features})
+        self.add_item(name, {"plain": {"authJSON": auth_json}}, **common_features)
```
Comment (azgabur) [inline testsuite/openshift/objects/auth_config/sections.py:160]: Small bugfix

File: testsuite/objects/__init__.py [hunk 10]
```diff
@@ -1,10 +1,10 @@
 """Module containing base classes for common objects"""
 import abc
-from dataclasses import dataclass, is_dataclass, fields
+from dataclasses import dataclass, is_dataclass, fields, field
 from copy import deepcopy
-from typing import Literal, Union
+from typing import Literal, Optional
 
-JSONValues = Union[None, str, int, bool, list["JSONValues"], dict[str, "JSONValues"]]
+JSONValues = None | str | int | bool | list["JSONValues"] | dict[str, "JSONValues"]
```
Comment (azgabur) [inline testsuite/objects/__init__.py:7]: Removing union in favor of new definition

File: testsuite/objects/__init__.py [hunk 11]
```diff
@@ -86,15 +93,21 @@ class Value(ABCValue):
     value: JSONValues
 
 
+@dataclass
+class _AuthJson:
```
Comment (mganisin) [inline testsuite/objects/__init__.py:97]: I don't understand this, a brief note in commit message would help (I am sorry if it is already there), it has no other use than in ValueFrom, so the value/benefit is unclear

Also I suggest in these cases to follow more strictly naming scheme of underlying objects as much as possible/reasonable. So I would keep capital 'A' however I would keep JSON capitalized so 'AuthJSON' is in my opinion best option.

Why 'private' with '_'?

File: testsuite/objects/__init__.py [hunk 12]
```diff
@@ -86,15 +93,21 @@ class Value(ABCValue):
...
+    authJSON: str  # pylint: disable=invalid-name
+
+
 @dataclass
 class ValueFrom(ABCValue):
     """Dataclass for dynamic Value. It contains reference path to existing value in AuthJson."""
 
-    authJSON: str  # pylint: disable=invalid-name
+    valueFrom: _AuthJson | str  # pylint: disable=invalid-name
```
Comment (mganisin) [inline testsuite/objects/__init__.py:105]: And if so, why keeping 'str' and '__post_init__'? (I see why, however the tradeoff to remove it doesn't seem to be that big)

File: testsuite/tests/kuadrant/authorino/identity/extended_properties/test_extended_properties.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    authorization.responses.add_simple("auth.identity")
+    return authorization
+
+
+def test_basic(client, auth):
+    """
+    This test checks if static and dynamic extended properties are created and have the right value.
+    """
+    response = client.get("/anything/abc", auth=auth)
+    assert extract_response(response)["property_static"] % "" == "static"
```
Comment (mganisin) [inline testsuite/tests/kuadrant/authorino/identity/extended_properties/test_extended_properties.py:37]: "NOT_SET", "MISSING" or any label o this kind would be bit more helpful during test results review than empty string

**Comment by mganisin** [comment]
Removal of `Union` and use of `kw_only` deserves special acknowledgement :-)

File: testsuite/objects/__init__.py (hunk 11 above)
Comment (azgabur) [reply testsuite/objects/__init__.py]: I renamed the class and unmarked it as private. 
I was at first opposing creation of new class just to solve the custom asdict function, but after extending functionality of `ABCValue` I did not find a way to make the custom asdict function to work so I resorted to creating this new class to solve the problem. Using `__post_init__` to transform `str` to instance of `AuthJSON` class makes using the class still straightforward.
For example
```py
from from testsuite.objects import ValueFrom

my_value = ValueFrom("somewhere.here")
```
instead of:
```py
from from testsuite.objects import ValueFrom, AuthJSON

my_value = ValueFrom(AuthJSON("somewhere.here"))
```
But in the end both ways will work in current code.

I made it private at first because I did not see the second usage in my example as very real.

File: testsuite/tests/kuadrant/authorino/identity/extended_properties/test_extended_properties.py (hunk 13 above)
Comment (azgabur) [reply testsuite/tests/kuadrant/authorino/identity/extended_properties/test_extended_properties.py]: Good idea!

File: testsuite/objects/__init__.py (hunk 11 above)
Comment (azgabur) [reply testsuite/objects/__init__.py]: In my recent push I remade the the logic once more. I removed the new class and used [Init-only variable](https://docs.python.org/3/library/dataclasses.html#init-only-variables) and subsequent transformation in `__post_init__`

File: testsuite/objects/__init__.py [hunk 14]
```diff
@@ -76,8 +76,19 @@ class ABCValue(abc.ABC):
...
+        - name: for use as JsonProperty
+        - overwrite: for use in ExtendedProperty
+
+    The kw_only argument is used to avoid the need to define default variables in child classes which would otherwise
+    be needed due to how dataclass inheritance works. As these variables are optional it is natural to make them be
+    used as keyword arguments instead of positional arguments.
     """
 
+    name: Optional[str] = field(default=None, kw_only=True)
+    overwrite: Optional[bool] = field(default=None, kw_only=True)
```
Comment (pehala) [inline testsuite/objects/__init__.py:90]: Having this in an abstract class just for one implementation seems wrong

Comment (azgabur) [reply testsuite/objects/__init__.py]: Moved to logic to extra classes as I had previously. I later realized it is a bad idea.

File: testsuite/objects/__init__.py [hunk 15]
```diff
@@ -1,17 +1,17 @@
...
-JSONValues = Union[None, str, int, bool, list["JSONValues"], dict[str, "JSONValues"]]
+JSONValues = None | str | int | bool | list["JSONValues"] | dict[str, "JSONValues"]
 
 
 def asdict(obj) -> dict[str, JSONValues]:
     """
     This function converts dataclass object to dictionary.
     While it works similar to `dataclasses.asdict` a notable change is usage of
-    overriding `to_dict()` function if dataclass contains it.
+    overriding `asdict()` function if dataclass contains it.
```
Comment (azgabur) [inline testsuite/objects/__init__.py:14]: fix typo

File: testsuite/objects/__init__.py (hunk 11 above)
Comment (azgabur) [reply testsuite/objects/__init__.py]: I reverted to original. Will keep the thought for future.

File: testsuite/objects/__init__.py [hunk 16]
```diff
@@ -109,6 +109,22 @@ def asdict(self):
...
+    """
+    Dataclass extending Property class adding optional `overwrite` feature
+    used in extended_properties functionality in Identity section.
+    """
+
+    overwrite: Optional[bool] = None
+
+    def asdict(self):
+        """Override `asdict` function."""
+        if self.overwrite is not None:
```
Comment (pehala) [inline testsuite/objects/__init__.py:123]: Shouldnt custom asdict take care of this?

Comment (azgabur) [reply testsuite/objects/__init__.py:123]: I inherit from a class that defines `asdict()` override method so I need to redefine it, otherwise the parent method would be used.
But yes, the custom `asdict()` automatically do not include None variables.

File: testsuite/objects/__init__.py [hunk 17]
```diff
@@ -109,6 +109,22 @@ def asdict(self):
...
+class ExtendedProperty(Property):
+    """
+    Dataclass extending Property class adding optional `overwrite` feature
+    used in extended_properties functionality in Identity section.
+    """
+
+    overwrite: Optional[bool] = None
+
+    def asdict(self):
+        """Override `asdict` function."""
```
Comment (mganisin) [inline testsuite/objects/__init__.py:122]: Or maybe something like:
```
"""Extending inherited asdict() to include also new attribute(s) defined here"""
```

(btw this means, some sort of generaral asdict method that covers attributes generally might be handy; something to investigate in future)

Comment (azgabur) [reply testsuite/objects/__init__.py]: Changed the comment.
Your review gave me idea for a new function (#230). It wont solve this particular inheritance problem but a different one.

**Review by pehala — APPROVED** [review]

**Comment by azgabur** [comment]
^Fix merge conflict & rebase

File: testsuite/openshift/objects/auth_config/sections.py [hunk 18]
```diff
@@ -61,6 +62,23 @@ def add_item(
 class Identities(Section):
     """Section which contains identity configuration"""
 
+    def add_item(
```
[the remark this answers is not in the document]
Comment (azgabur) [reply testsuite/openshift/objects/auth_config/sections.py]: I kept the overwrite, but by using the star '*' syntax I overcame the problem of changing order of function arguments, that allowed me to simply add `extended_properties` without the need to list every parent argument.
I changed the parent function to facilitate this as well: https://github.com/Kuadrant/testsuite/pull/212#discussion_r1305881457

**Review by averevki — APPROVED** [review]

**Review by pehala — APPROVED** [review]
